lib/github_classroom/lti/membership_service.rb

```diff
@@ -0,0 +1,63 @@
+# frozen_string_literal: true
+
+module GitHubClassroom
+  module LTI
+    class MembershipService
+      include RequestSigning
+
+      def initialize(endpoint, consumer_key, secret)
+        @endpoint = endpoint
+        @consumer_key = consumer_key
+        @secret = secret
+      end
+
+      def students
+        filtered_membership(roles: %w[Student Learner])
+      end
+
+      def instructors
+        filtered_membership(roles: ["Instructor"])
+      end
+
+      def membership
+        req_headers = { "Accept": "application/vnd.ims.lis.v2.membershipcontainer+json" }
+        response = signed_request(@endpoint, @consumer_key, @secret, headers: req_headers).get
+
+        json_membership = JSON.parse(response.body)
+        parsed_membership = parse_membership(json_membership)
+
+        parsed_membership
+      end
+
+      private
+
+      def filtered_membership(roles: nil)
+        unfiltered_membership = membership
+        return unfiltered_membership unless roles
+
+        unfiltered_membership.select do |membership|
+          # LMS is allowed to prefix a role with anything it wants
+          # so the filter method simply matches on substrings
+          roles.any? do |desired_role|
+            membership.role.any? do |actual_role|
+              actual_role.include? desired_role
```

Comment:
What will `roles` and `unfiltered_membership` look like here?  We have 4 levels of iteration here, can we improve this at all? If these are long lists, this method could be extremely slow.

Reply:
I agree that this method is... unideal :grimace:

`Roles` is a list of strings. LTI defines 8 standard roles (Student, Admin, Learner, Instructor, etc) that may or may not have custom prefixes, but also allows an LMS to define any custom roles they want. In my limited experience testing and researching I haven’t seen a list of roles longer than 2 elements.

`unfiltered_membership` is just a collection of all the members associated with a given course, so I’d imagine depending on the university it could get pretty large, but not more than a few hundred?

Reply:
Thats awesome 😄 Please ping me again when this is ready for review again!